--- /u/j/y/jyoon/EC2.pm	Thu May  8 15:43:26 2008
+++ EC2.pm	Thu May  8 15:36:40 2008
@@ -99,7 +99,8 @@ sub _init {
 	my $ts = time2isoz();
 
 	$self->{signature_version} = 1;
-	$self->{version} = '2007-01-19';
+	#$self->{version} = '2007-01-19';
+	$self->{version} = '2007-08-29';
 	$self->{base_url} = 'http://ec2.amazonaws.com';
 	$self->{AWSAccessKeyId} = $args{AWSAccessKeyId};
 	$self->{SecretAccessKey} = $args{SecretAccessKey};
@@ -209,6 +210,7 @@ sub register_image {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -295,6 +297,7 @@ sub describe_images {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 		
 		return undef;
 	}
@@ -331,6 +334,7 @@ sub deregister_image {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -411,6 +415,7 @@ sub run_instances {
 								KeyName			=> { type => SCALAR, optional => 1 },
 								SecurityGroup	=> { type => SCALAR | ARRAYREF, optional => 1 },
 								UserData		=> { type => SCALAR, optional => 1 },
+								InstanceType	=> { type => SCALAR, optional => 1 },
 								AddressingType	=> { type => SCALAR, optional => 1 },
 	});
 	
@@ -429,6 +434,7 @@ sub run_instances {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -506,6 +512,7 @@ sub describe_instances {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -583,6 +590,7 @@ sub terminate_instances {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -640,6 +648,7 @@ sub create_key_pair {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -694,6 +703,7 @@ sub describe_key_pairs {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -733,6 +743,7 @@ sub delete_key_pair {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -795,6 +806,7 @@ sub modify_image_attribute {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -835,6 +847,7 @@ sub reset_image_attribute {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -876,6 +889,7 @@ sub create_security_group {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -947,6 +961,7 @@ sub describe_security_groups {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -983,6 +998,7 @@ sub delete_security_group {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -1059,6 +1075,7 @@ sub authorize_security_group_ingress {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -1135,6 +1152,7 @@ sub revoke_security_group_ingress {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -1171,6 +1189,7 @@ sub get_console_output {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -1216,6 +1235,7 @@ sub reboot_instances {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -1266,6 +1286,7 @@ sub describe_image_attribute {
 	if ( grep { defined && length } $xml->{Errors} ) {
 		$self->_debug("ERROR: $xml->{Errors}{Error}{Message}");
 		$self->{error} = $xml->{Errors}{Error}{Message};
+		$self->{errorcode} = $xml->{Errors}{Error}{Code};
 
 		return undef;
 	}
@@ -1306,4 +1327,4 @@ under the same terms as Perl itself.
 
 =head1 SEE ALSO
 
-Amazon EC2 API: L<http://docs.amazonwebservices.com/AmazonEC2/dg/2007-01-19/>
\ No newline at end of file
+Amazon EC2 API: L<http://docs.amazonwebservices.com/AmazonEC2/dg/2007-01-19/>
